backend/btrixcloud/colls.py

```diff
@@ -391,10 +581,228 @@ async def add_successful_crawl_to_collections(self, crawl_id: str, cid: UUID):
+                count=data.get("count", 0),
+                snapshots=[
+                    PageIdTimestamp(
+                        pageId=p["_id"], ts=p.get("ts"), status=p.get("status", 200)
+                    )
+                    for p in data.get("pages", [])
+                ],
+            )
+            for data in items
+        ], total
+
+    async def set_home_url(
+        self, coll_id: UUID, update: UpdateCollHomeUrl, org: Organization
+    ) -> Dict[str, bool]:
+        """Set home URL for collection and save thumbnail to database"""
+        if update.pageId:
+            page = await self.page_ops.get_page(update.pageId, org.id)
+            update_query = {
+                "homeUrl": page.url,
+                "homeUrlTs": page.ts,
+                "homeUrlPageId": page.id,
+            }
+        else:
+            update_query = {
+                "homeUrl": None,
+                "homeUrlTs": None,
+                "homeUrlPageId": None,
+            }
+
+        await self.collections.find_one_and_update(
+            {"_id": coll_id, "oid": org.id},
+            {"$set": update_query},
+        )
+
+        return {"updated": True}
+
+    async def upload_thumbnail_stream(
+        self, stream, filename: str, coll_id: UUID, org: Organization, user: User
+    ) -> Dict[str, bool]:
+        """Upload file as stream to use as collection thumbnail"""
+        coll = await self.get_collection(coll_id)
+
+        _, extension = os.path.splitext(filename)
+
+        image_filename = f"thumbnail-{str(coll_id)}{extension}"
+
+        prefix = org.storage.get_storage_extra_path(str(org.id)) + "images/"
+
+        file_prep = ImageFilePreparer(
+            prefix,
+            image_filename,
+            original_filename=filename,
+            user=user,
+            created=dt_now(),
+        )
+
+        async def stream_iter():
+            """iterate over each chunk and compute and digest + total size"""
+            async for chunk in stream:
+                file_prep.add_chunk(chunk)
+                yield chunk
+
+        print("Collection thumbnail stream upload starting", flush=True)
+
+        if not await self.storage_ops.do_upload_multipart(
+            org,
+            file_prep.upload_name,
+            stream_iter(),
+            MIN_UPLOAD_PART_SIZE,
+        ):
+            print("Collection thumbnail stream upload failed", flush=True)
+            raise HTTPException(status_code=400, detail="upload_failed")
+
+        print("Collection thumbnail stream upload complete", flush=True)
+
+        thumbnail_file = file_prep.get_image_file(org.storage)
+
+        if thumbnail_file.size > THUMBNAIL_MAX_SIZE:
+            print(
+                "Collection thumbnail stream upload failed: max size (2 MB) exceeded",
+                flush=True,
+            )
+            await self.storage_ops.delete_file_object(org, thumbnail_file)
+            raise HTTPException(status_code=400, detail="upload_failed")
```

Comment:
Should we pass this on as a more descriptive error, like `{"error": max_thumbnail_size_exceeded, "max_size": THUMBNAIL_MAX_SIZE}`